partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,893 @@
+    }
+
+    #[test]
+    fn simple_sfw_with_alias() {
+        // Open Schema with `Strict` typing mode and `age` in nested attribute.
+        let details_fields = struct_fields![("age", int!())];
+        let details = r#struct![BTreeSet::from([details_fields])];
+
+        assert_query_typing(
+            TypingMode::Strict,
+            "SELECT d.age FROM customers.details AS d",
+            create_customer_schema(
+                true,
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("details", details.clone()),
+                ],
+            ),
+            vec![StructField::new("age", int!())],
+        )
+        .expect("Type");
+
+        assert_query_typing(
+            TypingMode::Strict,
+            "SELECT c.id AS my_id, customers.name AS my_name FROM customers AS c",
+            create_customer_schema(
+                false,
+                vec![
+                    StructField::new("id", int!()),
+                    StructField::new("name", str!()),
+                    StructField::new("age", any!()),
+                ],
+            ),
+            vec![
+                StructField::new("my_id", int!()),
+                StructField::new("my_name", str!()),
+            ],
+        )
+        .expect("Type");
+    }
+
+    #[test]
+    fn simple_sfw_err() {
+        // Closed Schema with `Strict` typing mode and `age` non-existent projection.
+        assert_err(
+            assert_query_typing(
+                TypingMode::Strict,
+                "SELECT customers.id, customers.name, customers.age FROM customers",
+                create_customer_schema(
+                    false,
+                    vec![
+                        StructField::new("id", int!()),
+                        StructField::new("name", str!()),
+                    ],
+                ),
+                vec![],
+            ),
+            vec![TypingError::IllegalState(format!(
+                "Illegal Derive Type PartiqlType(String)"
+            ))],
+        );
+
+        // Closed Schema with `Strict` typing mode and `bar` non-existent projection from closed nested `details`.
+        let details_fields = struct_fields![("age", int!())];
+        let details = r#struct![BTreeSet::from([
+            details_fields,
+            StructConstraint::Open(false)
+        ])];
+        assert_err(
+            assert_query_typing(
+                TypingMode::Strict,
+                "SELECT customers.id, customers.name, customers.details.bar FROM customers",
+                create_customer_schema(
+                    false,
+                    vec![
+                        StructField::new("id", int!()),
+                        StructField::new("name", str!()),
+                        StructField::new("details", details),
+                    ],
+                ),
+                vec![],
```

Comment:
In the error case + strict mode, should we still emit some type info? Like for instance, stating `id` resolved to `int`, `name` resolved to `string`, and `details.bar` resolved to `undefined`?

Reply:
I don't believe we should, as it is ill-typed. However we can add the last type in the stack to the error message which should be the `output_schema`.